src/components/DataDimension/DataTypeSelector.js

```diff
@@ -2,57 +2,73 @@ import { SingleSelectField, SingleSelectOption } from '@dhis2/ui'
 import PropTypes from 'prop-types'
 import React from 'react'
 import i18n from '../../locales/index.js'
-import {
-    DIMENSION_TYPE_ALL,
-    dataTypeMap as dataTypes,
-    DIMENSION_TYPE_EXPRESSION_DIMENSION_ITEM,
-} from '../../modules/dataTypes.js'
+import { DIMENSION_TYPE_ALL, dataTypeMap } from '../../modules/dataTypes.js'
+import { getDisplayNameByVisType } from '../../modules/visTypes.js'
+import { useDataDimensionContext } from './DataDimension.js'
 import styles from './styles/DataTypeSelector.style.js'
 
 const DataTypeSelector = ({
     currentDataType,
+    dataTypes,
     onChange,
     dataTest,
-    includeCalculations,
-}) => (
-    <div className="container">
-        <SingleSelectField
-            label={i18n.t('Data Type')}
-            dataTest={dataTest}
-            selected={dataTypes[currentDataType]?.id || DIMENSION_TYPE_ALL}
-            onChange={(ref) => onChange(ref.selected)}
-            dense
-        >
-            <SingleSelectOption
-                value={DIMENSION_TYPE_ALL}
-                key={DIMENSION_TYPE_ALL}
-                label={i18n.t('All types')}
-                dataTest={`${dataTest}-option-all`}
-            />
-            {Object.values(dataTypes)
-                .filter(
-                    (type) =>
-                        type.id !== DIMENSION_TYPE_EXPRESSION_DIMENSION_ITEM ||
-                        includeCalculations
-                )
-                .map((type) => (
+}) => {
+    const { visType } = useDataDimensionContext()
+
+    return (
```

Comment:
There are clearly two different flows here, when `dataTypes.length === 1` (new) and when it isn't (old). For clarity, maybe it's better to make the distinction early to avoid having to do the checks in so many places. So I would prefer to refactor this to something like `return dataTypes.length === 1 ? [ SingleSelectField for NEW ] : [ SingleSelectField for OLD ]`

Reply:
if you then want to avoid some duplication, like `label={i18n.t('Data Type')}`, that can be extracted out to a const like `const label ={i18n.t('Data Type')}` to be used in both `SingleSelectFields` instead

Reply:
I tried this and had issues with re-rendering of components and errors due to the select trying to preselect an item that is not present in the list...

Reply:
The selected item must be available to select, yes. But that sounds like you need to change what goes in to the `selected` prop, maybe add a condition there to make sure it's passed properly?

Reply:
I tried again, I must have done something wrong the previous time, not it seems to work fine.

Reply:
This is much cleaner and easier to read, great improvement 👍